pandas/core/computation/parsing.py

```diff
@@ -168,6 +179,93 @@ def tokenize_backtick_quoted_string(
     return BACKTICK_QUOTED_STRING, source[string_start:string_end]
 
 
+def _split_by_backtick(s: str) -> list[tuple[bool, str]]:
+    """
+    Splits a str into substrings along backtick characters (`).
+
+    Disregards backticks inside quotes.
+
+    Parameters
+    ----------
+    s : str
+        The Python source code string.
+
+    Returns
+    -------
+    substrings: list[tuple[bool, str]]
+        List of tuples, where each tuple has two elements:
+        The first is a boolean indicating if the substring is backtick-quoted.
+        The second is the actual substring.
+    """
+    substrings = []
+    substring = ""
+    i = 0
+    while i < len(s):
+        backtick_index = s.find("`", i)
+
+        # No backticks
```

Comment:
I definitely appreciate all the comments that have been added here, but generally I feel like this would be better modeled by managing the parsing state. Something to the effect of:

```python
class ParseState(Enum):
    DEFAULT = 0
    IN_BACKTICK = 1
    IN_SINGLE_QUOTE = 2
    IN_DOUBLE_QUOTE = 3
    # below may or may not be required, depending on implementation
    CLOSING_BACKTICK = 4
    CLOSING_SINGLE_QUOTE = 5
    CLOSING_DOUBLE_QUOTE = 6
```

Then you can start collecting the substrings until the state or string should end:

```python
state = ParseState.DEFAULT
substrings = []
for substr in s:
    match substr:
        case "'":
            if state == ParseState.DEFAULT:
                state = ParseState.IN_SINGLE_QUOTE
            else:
                substrings.append(substr)
         case "\"":
             if state == ParseState.DEFAULT:
                 state = ParseState.IN_DOUBLE_QUOTE
             else:
                 substrings.append(substr)
```

There's a lot of logic that needs to be filled in but I think that will help you lay this out better than assigning -1 to different variables

Reply:
1. Hm I'm not sure how many `-1`'s I can get rid of, since it's a return value of `s.find()` that needs to be accounted for.
2. I assign `-1` to two variables here, and it's consistent with the meaning of `-1` returned from `s.find()` i.e. the character wasn't found in the string and therefore has no nonnegative index.
3. I'm not going character by character, instead I'm finding the opening quote/backtick and then the closing quote/backtick

Did you have in mind something like this? (caveat: Repeated code isn't cleaned up yet, etc. etc.)
```python
    substrings = []
    substring = ""
    i = 0
    while i < len(s):
        backtick_index = s.find("`", i)

        if backtick_index == -1:
            parse_state = ParseState.NO_BACKTICK

        else:
            single_quote_index = s.find("'", i)
            double_quote_index = s.find('"', i)
            quote_index = (
                -1 if (single_quote_index == -1) and (double_quote_index == -1)
                else double_quote_index if single_quote_index == -1
                else single_quote_index if double_quote_index == -1
                else min(double_quote_index, single_quote_index)
            )

            # No quotes, or
            # Backtick opened before quote
            if (quote_index == -1) or (backtick_index < quote_index):
                next_backtick_index = s.find("`", backtick_index + 1)
                while (
                    (next_backtick_index != -1)
                    and (next_backtick_index != len(s) - 1)
                    and (s[next_backtick_index + 1] == "`")
                ):
                    # Since the next character is also a backtick, it's an escaped backtick
                    next_backtick_index = s.find("`", next_backtick_index + 2)

                if next_backtick_index == -1:
                    parse_state = ParseState.BACKTICK_UNMATCHED
                else:
                    parse_state = ParseState.BACKTICK_MATCHED

            # Quote opened before backtick
            else:
                next_quote_index = -1
                line_reader = BytesIO(s[i:].encode("utf-8")).readline
                token_generator = tokenize.tokenize(line_reader)
                for toknum, _, (_, _), (_, end), _ in token_generator:
                    if toknum == tokenize.STRING:
                        next_quote_index = i + end - 1
                        break

                if next_quote_index == -1:
                    parse_state = ParseState.QUOTE_UNMATCHED
                elif next_quote_index == len(s) - 1:
                    parse_state = ParseState.QUOTE_MATCHED_AT_THE_END
                else:
                    parse_state = ParseState.QUOTE_MATCHED_IN_THE_MIDDLE

        match parse_state:
            case ParseState.NO_BACKTICK:
                substrings.append((False, substring + s[i:]))
                break

            case ParseState.BACKTICK_UNMATCHED:
                substrings.append((False, substring + s[i:]))
                break

            case ParseState.BACKTICK_MATCHED:
                substring += s[i:backtick_index]
                if substring:
                    substrings.append((False, substring))
                substrings.append((True, s[backtick_index : next_backtick_index + 1]))
                substring = ""
                i = next_backtick_index + 1
                continue

            case ParseState.QUOTE_UNMATCHED:
                substrings.append((False, substring + s[i:]))
                break

            case ParseState.QUOTE_MATCHED_AT_THE_END:
                substrings.append((False, substring + s[i:]))
                break

            case ParseState.QUOTE_MATCHED_IN_THE_MIDDLE:
                substring += s[i : next_quote_index + 1]
                i = next_quote_index + 1
                continue

    return substrings
```

Reply:
Kind of, although I was hoping all of the actions that needed to be taken could be handled within the state branch. For example, with the backticks, instead of using .find up front, I think you could do:

```python
strlen = len(s)
while i < strlen:
    match substr:
        case "`":
            if state == ParseState.DEFAULT:
                state = ParseState.BACKTICK_MATCHED
                break
            elif state == ParseState.BACKTICK_MATCHED:
                if i == strlen:
                    break
                elif substring[i + 1] == "`":
                    substrings.append(substr)
                    i += 1;  # can advance past next backtick
                    break
                else:
                    # I'm assuming the string should start/end with a backtick
                    # may not be the right assumption, so adjust logic accordingly
                    raise ValueError("unable to parse")
            else:
                substrings.append(substr)
                
        ...
        
        case _:
            substrings.append(substr)
    i += 1
```